src/GafferScene/Capsule.cpp

```diff
@@ -84,65 +96,15 @@ Capsule::Capsule(
 
 Capsule::~Capsule()
 {
-	// Disconnect from signals
-	setScene( nullptr );
 }
 
 void Capsule::setScene( const ScenePlug *scene )
 {
-	assert( !scene || scene->parent() );
-
-	// Connecting to and disconnecting from signals is not threadsafe,
-	// and neither is calling `GraphComponent::parentChangedSignal()`
-	// (because it constructs signals lazily on demand). Most (all?)
-	// other signal access occurs on the UI thread, but capsules are
-	// constructed concurrently on background threads when generating
-	// scenes. We use a mutex to serialise all signal access performed
-	// by capsules, but this is still not safe with respect to signal
-	// access that may be performed by UI components on the main thread.
-	//
-	// Todo : Make this watertight. Possibilities include :
-	//
-	// - Switching to boost::signals2, which is threadsafe by default,
-	//   and making the GraphComponent signal accessors threadsafe.
-	// - Writing our own threadsafe signals classes which do exactly what
-	//   we need, without the bloat of the boost versions. And making the
-	//   GraphComponent signal accessors threadsafe.
-	// - Ditching all the signal handling in capsules. It is only used
-	//   to track the erroneous usage of "expired" capsules, which can
-	//   only arise from bugs elsewhere. I can't recall seeing an expired
-	//   capsule yet.
-	static tbb::spin_mutex g_signalMutex;
-	tbb::spin_mutex::scoped_lock signalLock( g_signalMutex );
-
-	if( const Node *node = m_scene ? m_scene->node() : nullptr )
-	{
-		const_cast<Node *>( node )->plugDirtiedSignal().disconnect(
-			boost::bind( &Capsule::plugDirtied, this, ::_1 )
-		);
-	}
-	if( m_scene )
-	{
-		const_cast<ScenePlug *>( m_scene )->parentChangedSignal().disconnect(
-			boost::bind( &Capsule::parentChanged, this, ::_1 )
-		);
-	}
-
 	m_scene = scene;
-
 	if( m_scene )
 	{
-		const_cast<ScenePlug *>( m_scene )->parentChangedSignal().connect(
-			boost::bind( &Capsule::parentChanged, this, ::_1 )
-		);
```

Comment:
`m_scene` is a raw pointer, and there's a comment in the header saying that we don't need to own a reference because we're using `parentChanged()` to invalidate it instead. Do we need to own a reference now, or should we leave the parent tracking in place for now? Would be nice to get rid of all the signal handling so maybe the reference is the way to go? The choice of raw pointers seems a little odd in the first place, so I'm wondering if my old self was aware of something that I'm not. Hopefully not. If we decide to own a reference, we should also change the constructor argument to `const ConstScenePlugPtr &` to communicate the semantics.

Reply:
Hmm, I got a bit stumped by this one.  Doing an obvious version of this produces a segfault during exit while destructing the LRUCache.  Is there a limited set of circumstances where it's legal to destruct a ScenePlug?  I guess I'll leave this as a separate commit and see if you can figure it out.  Stack trace looks like this:
```#0  0x00007fffde2b0e28 in tbb::interface6::enumerable_thread_specific<Gaffer::Plug::DirtyPlugs, tbb::cache_aligned_allocator<Gaffer::Plug::DirtyPlugs>, (tbb::ets_key_usage_type)1>::create_local() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#1  0x00007fffde2b1052 in Gaffer::Plug::DirtyPlugs::local() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#2  0x00007fffde2ac9b9 in Gaffer::Plug::pushDirtyPropagationScope() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#3  0x00007fffde2666ac in Gaffer::GraphComponent::~GraphComponent() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#4  0x00007fffde2ade20 in Gaffer::Plug::~Plug() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#5  0x00007fffde30283e in Gaffer::ValuePlug::~ValuePlug() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#6  0x00007fffc316dc69 in GafferScene::ScenePlug::~ScenePlug() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGafferScene.so
#7  0x00007fffc306b5de in GafferScene::Capsule::~Capsule() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGafferScene.so
#8  0x00007fffc306b609 in GafferScene::Capsule::~Capsule() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGafferScene.so
#9  0x00007fffde303572 in IECorePreview::LRUCache<IECore::MurmurHash, boost::intrusive_ptr<IECore::Object const>, IECorePreview::LRUCachePolicy::TaskParallel, (anonymous namespace)::ComputeProcessKey>::~LRUCache() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#10 0x00007ffff6ae2b69 in __run_exit_handlers () from /lib64/libc.so.6
#11 0x00007ffff6ae2bb7 in exit () from /lib64/libc.so.6
#12 0x00007ffff78c8e1f in Py_Exit (sts=sts@entry=0) at Python/pythonrun.c:1774```

Reply:
I think we're just looking at a classic static destruction order fiasco. The cache is static, and so is the `g_dirtyPlugs` used to track plug dirtiness. `g_dirtyPlugs` has been destroyed already, and when the cache is destroyed it deletes a plug which then tries to access `g_dirtyPlugs`. I think our options are to stick with the raw pointer and parent tracking, or to do a tactical "leak" of `g_dirtyPlugs` by declaring it as `static tbb::enumerable_thread_specific<Plug::DirtyPlugs> *g_dirtyPlugs = new ...`;

What we really want is weak pointer support in IECore::RefCounted, or a derived class of it. I've been thinking a bit about adding that as I think there are a few use cases for it, but that'll be a longer term thing.